_posts/2023-11-16-quarkus-meets-langchain4j.adoc

```diff
@@ -0,0 +1,236 @@
+
+    Respond with a JSON document containing:
+    - the 'evaluation' key set to 'POSITIVE' if the review is positive, 'NEGATIVE' otherwise
+    - the 'message' key set to a message thanking or apologizing to the customer. These messages must be polite and match the review's language.
+
+    ---
+    {review}
+    ---
+""")
+TriagedReview triage(String review);
+----
+
+Voilà! That's all you need to do to describe the interaction with the LLM. The instructions follow a set of principles to shape the LLM response. Learn more about these techniques in [the dedicated prompt engineering page](https://docs.quarkiverse.io/quarkus-langchain4j/dev/prompt-engineering.html).
+
+Now, to call the LLM from the application code, just inject the `TriageService` and call the `triage` method:
+
+[source, java]
+----
+@Path("/review")
+public class ReviewResource {
+
+    @Inject
+    TriageService triage;
+
+    record Review(String review) {
+      // User text
+    }
+
+    @POST
+    public TriagedReview triage(Review review) {
+        return triage.triage(review.review());
+    }
+
+}
+----
+
+That's it! The LLM is now integrated into the application. The `TriageService` interface is used as an ambassador to call the LLM. This declarative approach has many advantages:
+
+- Testability - you can easily mock the LLM by providing a fake implementation of the interface.
+- Observability - you can use the Quarkus metrics annotation to monitor the LLM methods.
+- Resilience - you can use the Quarkus fault-tolerance annotations to handle failures, timeouts, and other transient issues.
+
+== Tools and Document loader
+
+The previous example is a bit simplistic. In the real world, you will need to extend the LLM knowledge with tools and document stores. The `@RegisterAiService` annotation lets you define the tools and document stores to use.
+
+=== Tools
+
+Tools are methods that the LLM can invoke.
+
+To declare a tool, just use the `@Tool` annotation in a bean method:
+
+[source, java]
+----
+@ApplicationScoped
+public class CustomerRepository implements PanacheRepository<Customer> {
+
+    @Tool("get the customer name for the given customerId")
+    public String getCustomerName(long id) {
+        return find("id", id).firstResult().name;
+    }
+
+}
+----
+
+In this example, we are using the Panache repository pattern to access the database. We have a specific method annotated with `@Tool` to retrieve the customer name. When the LLM needs to get the customer name, it instructs Quarkus to call this method and receives the result.
+
+Obviously, it's not a good idea to expose every operation to the LLM. So, in addition to `@Tool`, you need to list the set of tools you allow the LLM to invoke in the `@RegisterAiService` annotation:
+
+[source, java]
+----
+@RegisterAiService(tools = { TransactionRepository.class, CustomerRepository.class })
+public interface FraudDetectionAi {
+   // ...
+}
+----
+
+ Note that not all LLMs support tools.
```

Comment:
should we mention somewhere which LLM's that are currently supported/working with quarkus-langchain4j?

Reply:
only openai unfortunately. I didn't find a HF model supporting tools yet (but only tried 5 of them)

Reply:
I don't know about ollama (@geoand do you know)?

Reply:
My guess is that it does not as the REST request / response does not contain anything indicating such a possibility